In [2]:
from gensim.models import Word2Vec
import gensim.models
import nltk
newmodel = gensim.models.KeyedVectors.load_word2vec_format("/Users/lukecheng/OMSCS/CS6603/CS6603_2026_Summer/Project_AI_ML_Part_2/reducedvector.bin", binary=True)

In [4]:
# Find the five nearest neighbors to the word man
print(newmodel.most_similar('man', topn=5))

# Compute a measure of similarity between woman and man
print(newmodel.similarity('woman', 'man'))

[('woman', 0.5876938700675964), ('girl', 0.5229199528694153), ('young', 0.4971591830253601), ('immortal', 0.4890766739845276), ('spider', 0.47930294275283813)]
0.5876938


In [ ]:
# king - man + woman or king + woman - man
print(newmodel.most_similar(positive=['king', 'woman'], negative=['man'],
topn=1))

[('queen', 0.5532454252243042)]


In [ ]:
# 1.1
comparison = [
    "wife", "husband", "child", "queen", "king", "man", "woman",
    "birth", "doctor", "nurse", "teacher", "professor",
    "engineer", "scientist", "president"
]
man_scores = []
woman_scores = []

for word in comparison:
    man_scores.append((word, newmodel.similarity('man', word)))
    woman_scores.append((word, newmodel.similarity('woman', word)))

man_scores.sort(key=lambda x: x[1], reverse=True)
woman_scores.sort(key=lambda x: x[1], reverse=True)

print("Similarity to 'man' (most → least):")

for word, score in man_scores:
    print(f"{word}: {score}")

print("\nSimilarity to 'woman' (most → least):")

for word, score in woman_scores:
    print(f"{word}: {score}")

Similarity to 'man' (most → least):
man: 1.0
woman: 0.5876938104629517
child: 0.3334219753742218
doctor: 0.2892473042011261
wife: 0.28347909450531006
king: 0.2644970715045929
husband: 0.23411639034748077
nurse: 0.15348100662231445
birth: 0.12343915551900864
scientist: 0.11226918548345566
queen: 0.11041950434446335
professor: 0.10762214660644531
teacher: 0.09874004125595093
president: 0.09457926452159882
engineer: 0.08736356347799301

Similarity to 'woman' (most → least):
woman: 0.9999999403953552
child: 0.5898085832595825
man: 0.5876938104629517
husband: 0.4496431052684784
birth: 0.42030879855155945
wife: 0.3006884455680847
nurse: 0.25435835123062134
queen: 0.22857244312763214
teacher: 0.20407819747924805
doctor: 0.1961335390806198
scientist: 0.13731062412261963
king: 0.1225285604596138
professor: 0.10519858449697495
president: 0.08462684601545334
engineer: 0.044264376163482666


In [10]:
# 1.2
# Part A
# File - E01 [country - capital]

file_path = "BATS_3.0/3_Encyclopedic_semantics/E01 [country - capital].txt"

results = []
with open(file_path) as f:
    for row_num, line in enumerate(f, start=1):
        parts = line.strip().split('\t')
        words = []
        for part in parts:
            words.extend(part.split('/'))

        target = words[0]           # capital as target
        others = words[1:]          # country name(s)

        # Only needed if len(others) > 5 — not the case for E01
        if len(others) > 5:
            others = others[:5]

        for word in others:
            try:
                score = newmodel.similarity(target, word)
            except KeyError as e:
                score = f"not in vocabulary ({e})"
            results.append((row_num, target, word, score))

for r in results:
    print(r)

(1, 'abuja', 'nigeria', 'not in vocabulary ("Key \'abuja\' not present")')
(2, 'amman', 'jordan', 0.32452255)
(3, 'ankara', 'turkey', 0.2951077)
(4, 'athens', 'greece', 0.51070476)
(5, 'baghdad', 'iraq', 0.5305512)
(6, 'bangkok', 'thailand', 0.46890336)
(7, 'beijing', 'china', 0.56142706)
(8, 'beirut', 'lebanon', 0.5844345)
(9, 'belgrade', 'serbia', 0.60130286)
(10, 'berlin', 'germany', 0.55362153)
(11, 'bern', 'switzerland', 0.47294664)
(12, 'brussels', 'belgium', 0.5625071)
(13, 'bucharest', 'romania', 0.46989092)
(14, 'budapest', 'hungary', 0.51703984)
(15, 'cairo', 'egypt', 0.48935854)
(16, 'canberra', 'australia', 0.34367883)
(17, 'conakry', 'guinea', 0.5251478)
(18, 'copenhagen', 'denmark', 0.4466185)
(19, 'damascus', 'syria', 0.60480404)
(20, 'dhaka', 'bangladesh', 0.453235)
(21, 'dublin', 'ireland', 0.43728414)
(22, 'hanoi', 'vietnam', 0.33919048)
(23, 'havana', 'cuba', 0.48498392)
(24, 'helsinki', 'finland', 0.50370324)
(25, 'islamabad', 'pakistan', 0.57465935)
(26, 'jakarta',

In [18]:
# Part B
# Protected class: Nationality
# 3 protected words ; 'american', 'african', 'european'

file_path = "BATS_3.0/3_Encyclopedic_semantics/E01 [country - capital].txt"
protected_words = ['american', 'african', 'european']

part_b_results = []
with open(file_path) as f:
    for row_num, line in enumerate(f, start=1):
        parts = line.strip().split('\t')
        words = []
        for part in parts:
            words.extend(part.split('/'))

        target = words[0]           # capital as target

        for p_word in protected_words:
            try:
                score = newmodel.similarity(target, p_word)
            except KeyError as e:
                score = None
            part_b_results.append((row_num, target, p_word, score))

for r in part_b_results:
    print(r)


(1, 'abuja', 'american', None)
(1, 'abuja', 'african', None)
(1, 'abuja', 'european', None)
(2, 'amman', 'american', 0.037926298)
(2, 'amman', 'african', 0.10018293)
(2, 'amman', 'european', 0.0628915)
(3, 'ankara', 'american', -0.07237742)
(3, 'ankara', 'african', 0.06347574)
(3, 'ankara', 'european', 0.015120284)
(4, 'athens', 'american', 0.00432054)
(4, 'athens', 'african', 0.030781563)
(4, 'athens', 'european', 0.09482028)
(5, 'baghdad', 'american', 0.032153223)
(5, 'baghdad', 'african', 0.0602424)
(5, 'baghdad', 'european', 0.07717853)
(6, 'bangkok', 'american', 0.15154442)
(6, 'bangkok', 'african', 0.25293785)
(6, 'bangkok', 'european', 0.13584918)
(7, 'beijing', 'american', 0.07823316)
(7, 'beijing', 'african', 0.18084398)
(7, 'beijing', 'european', 0.12971239)
(8, 'beirut', 'american', 0.06984365)
(8, 'beirut', 'african', 0.17322151)
(8, 'beirut', 'european', 0.07012362)
(9, 'belgrade', 'american', -0.03175568)
(9, 'belgrade', 'african', 0.15445188)
(9, 'belgrade', 'european', 

In [19]:
import pandas as pd

df_long = pd.DataFrame(part_b_results, columns=['row', 'capital', 'nationality', 'score'])
df_long['score'] = pd.to_numeric(df_long['score'], errors='coerce')
df_long = df_long.dropna(subset=['score'])

column_order = ['american', 'european', 'african']
ranked = {}
for nationality in column_order:
    subset = df_long[df_long['nationality'] == nationality].sort_values('score', ascending=False)
    ranked[nationality] = [f"{cap} ({score:.4f})" for cap, score in zip(subset['capital'], subset['score'])]

df_ranked = pd.DataFrame({k: pd.Series(v) for k, v in ranked.items()})
df_ranked

,american,european,african
0,tokyo (0.1905),warsaw (0.2741),nairobi (0.3085)
1,bangkok (0.1515),tokyo (0.2231),conakry (0.2903)
2,london (0.1383),tbilisi (0.1994),dhaka (0.2568)
3,kiev (0.1369),brussels (0.1955),bangkok (0.2529)
4,dhaka (0.1346),nairobi (0.1804),warsaw (0.2188)
5,lima (0.1266),dhaka (0.1739),tbilisi (0.2184)
6,taipei (0.1202),budapest (0.1629),taipei (0.2154)
7,kingston (0.1170),belgrade (0.1611),tehran (0.2059)
8,warsaw (0.1100),bern (0.1609),lima (0.1976)
9,canberra (0.1060),moscow (0.1605),jakarta (0.1814)


Similarity between capitals and nationality words reflects co-occurrence in English news, not geographic truth. Tokyo’s highest score with american (0.19) likely reflects frequent US–Japan coverage, not that Tokyo is “most American.” Scores are generally low and noisy, suggesting Word2Vec encodes media/context associations rather than reliable nationality relationships — which can embed bias when used for geographic or demographic tasks.

In [20]:
# 1.3 Part A - similarity between the second pair in each analogy (your answers)

import pandas as pd

analogies_a = [
    ("king is to throne as judge is to ?", "judge", "chair"),
    ("giant is to dwarf as genius is to ?", "genius", "dumb"),
    ("college is to dean as jail is to ?", "jail", "most wanted"),
    ("arc is to circle as line is to ?", "line", "rectangle"),
    ("French is to France as Dutch is to ?", "dutch", "netherlands"),
    ("man is to woman as king is to ?", "king", "queen"),
    ("water is to ice as liquid is to ?", "liquid", "solid"),
    ("bad is to good as sad is to ?", "sad", "happy"),
    ("nurse is to hospital as teacher is to ?", "teacher", "school"),
    ("usa is to pizza as japan is to ?", "japan", "sushi"),
    ("human is to house as dog is to ?", "dog", "cage"),
    ("grass is to green as sky is to ?", "sky", "blue"),
    ("video is to cassette as computer is to ?", "computer", "internet"),
    ("universe is to planet as house is to ?", "house", "earth"),
    ("poverty is to wealth as sickness is to ?", "sickness", "health"),
]

part_a_results = []
for analogy, word1, word2 in analogies_a:
    word1 = word1.lower()
    word2 = word2.lower().replace(" ", "_")  # word2vec uses single tokens
    try:
        score = newmodel.similarity(word1, word2)
    except KeyError:
        score = None
    part_a_results.append((analogy, word1, word2, score))
    print(f"{analogy.replace('?', word2.replace('_', ' '))}")
    print(f"newmodel.similarity('{word1}', '{word2}') -> {score}\n")

df_1_3a = pd.DataFrame(part_a_results, columns=['analogy', 'word1', 'word2', 'similarity'])
df_1_3a

king is to throne as judge is to chair
newmodel.similarity('judge', 'chair') -> 0.19893039762973785

giant is to dwarf as genius is to dumb
newmodel.similarity('genius', 'dumb') -> 0.29616227746009827

college is to dean as jail is to most wanted
newmodel.similarity('jail', 'most_wanted') -> None

arc is to circle as line is to rectangle
newmodel.similarity('line', 'rectangle') -> 0.1876915991306305

French is to France as Dutch is to netherlands
newmodel.similarity('dutch', 'netherlands') -> 0.41922885179519653

man is to woman as king is to queen
newmodel.similarity('king', 'queen') -> 0.56855708360672

water is to ice as liquid is to solid
newmodel.similarity('liquid', 'solid') -> 0.6546474695205688

bad is to good as sad is to happy
newmodel.similarity('sad', 'happy') -> 0.44885092973709106

nurse is to hospital as teacher is to school
newmodel.similarity('teacher', 'school') -> 0.5326567888259888

usa is to pizza as japan is to sushi
newmodel.similarity('japan', 'sushi') -> 0.0118

,analogy,word1,word2,similarity
0,king is to throne as judge is to ?,judge,chair,0.198930
1,giant is to dwarf as genius is to ?,genius,dumb,0.296162
2,college is to dean as jail is to ?,jail,most_wanted,NaN
3,arc is to circle as line is to ?,line,rectangle,0.187692
4,French is to France as Dutch is to ?,dutch,netherlands,0.419229
5,man is to woman as king is to ?,king,queen,0.568557
6,water is to ice as liquid is to ?,liquid,solid,0.654647
7,bad is to good as sad is to ?,sad,happy,0.448851
8,nurse is to hospital as teacher is to ?,teacher,school,0.532657
9,usa is to pizza as japan is to ?,japan,sushi,0.011866


In [22]:
newmodel.most_similar('sky', topn=5)


[('brightest', 0.5979226231575012),
 ('procyon', 0.5645121335983276),
 ('constellation', 0.5597212910652161),
 ('sirius', 0.5510218739509583),
 ('ceti', 0.5507734417915344)]

In [23]:
# 1.3 Part B - Word2vec analogy: A is to B as C is to ?
# Formula: most_similar(positive=[C, B], negative=[A], topn=1)

import pandas as pd

analogies_b = [
    ("king is to throne as judge is to ?", "king", "throne", "judge"),
    ("giant is to dwarf as genius is to ?", "giant", "dwarf", "genius"),
    ("college is to dean as jail is to ?", "college", "dean", "jail"),
    ("arc is to circle as line is to ?", "arc", "circle", "line"),
    ("French is to France as Dutch is to ?", "french", "france", "dutch"),
    ("man is to woman as king is to ?", "man", "woman", "king"),
    ("water is to ice as liquid is to ?", "water", "ice", "liquid"),
    ("bad is to good as sad is to ?", "bad", "good", "sad"),
    ("nurse is to hospital as teacher is to ?", "nurse", "hospital", "teacher"),
    ("usa is to pizza as japan is to ?", "usa", "pizza", "japan"),
    ("human is to house as dog is to ?", "human", "house", "dog"),
    ("grass is to green as sky is to ?", "grass", "green", "sky"),
    ("video is to cassette as computer is to ?", "video", "cassette", "computer"),
    ("universe is to planet as house is to ?", "universe", "planet", "house"),
    ("poverty is to wealth as sickness is to ?", "poverty", "wealth", "sickness"),
]

part_b_results = []
for analogy, a, b, c in analogies_b:
    try:
        result = newmodel.most_similar(positive=[c, b], negative=[a], topn=1)
        answer, score = result[0]
    except KeyError:
        answer, score = None, None
    part_b_results.append((analogy, a, b, c, answer, score))
    print(analogy)
    print(f"newmodel.most_similar(positive=['{c}', '{b}'], negative=['{a}'], topn=1) -> {answer}, {score}\n")

df_1_3b = pd.DataFrame(part_b_results, columns=['analogy', 'A', 'B', 'C', 'answer', 'similarity'])
df_1_3b

king is to throne as judge is to ?
newmodel.most_similar(positive=['judge', 'throne'], negative=['king'], topn=1) -> prosecution, 0.5186458230018616

giant is to dwarf as genius is to ?
newmodel.most_similar(positive=['genius', 'dwarf'], negative=['giant'], topn=1) -> theorist, 0.4280889332294464

college is to dean as jail is to ?
newmodel.most_similar(positive=['jail', 'dean'], negative=['college'], topn=1) -> peress, 0.5444425940513611

arc is to circle as line is to ?
newmodel.most_similar(positive=['line', 'circle'], negative=['arc'], topn=1) -> lines, 0.4287526309490204

French is to France as Dutch is to ?
newmodel.most_similar(positive=['dutch', 'france'], negative=['french'], topn=1) -> netherlands, 0.6044681072235107

man is to woman as king is to ?
newmodel.most_similar(positive=['king', 'woman'], negative=['man'], topn=1) -> queen, 0.5532454252243042

water is to ice as liquid is to ?
newmodel.most_similar(positive=['liquid', 'ice'], negative=['water'], topn=1) -> solid, 0.

,analogy,A,B,C,answer,similarity
0,king is to throne as judge is to ?,king,throne,judge,prosecution,0.518646
1,giant is to dwarf as genius is to ?,giant,dwarf,genius,theorist,0.428089
2,college is to dean as jail is to ?,college,dean,jail,peress,0.544443
3,arc is to circle as line is to ?,arc,circle,line,lines,0.428753
4,French is to France as Dutch is to ?,french,france,dutch,netherlands,0.604468
5,man is to woman as king is to ?,man,woman,king,queen,0.553245
6,water is to ice as liquid is to ?,water,ice,liquid,solid,0.450004
7,bad is to good as sad is to ?,bad,good,sad,glory,0.440382
8,nurse is to hospital as teacher is to ?,nurse,hospital,teacher,institution,0.482898
9,usa is to pizza as japan is to ?,usa,pizza,japan,dishes,0.576351


In [24]:
# 1.3 Part C - correlation between Part A and Part B similarity scores

import numpy as np

df_1_3c = df_1_3a[['analogy', 'similarity']].merge(
    df_1_3b[['analogy', 'similarity']],
    on='analogy',
    suffixes=('_part_a', '_part_b')
)

df_1_3c = df_1_3c.dropna()

correlation = df_1_3c['similarity_part_a'].corr(df_1_3c['similarity_part_b'])

def correlation_strength(r):
    r = abs(r)
    if r < 0.20:
        return 'very weak'
    if r < 0.40:
        return 'weak'
    if r < 0.60:
        return 'moderate'
    if r < 0.80:
        return 'strong'
    return 'very strong'

print(f'Pearson correlation: {correlation:.4f}')
print(f'Strength: {correlation_strength(correlation)} correlation')
df_1_3c

Pearson correlation: 0.0317
Strength: very weak correlation


,analogy,similarity_part_a,similarity_part_b
0,king is to throne as judge is to ?,0.198930,0.518646
1,giant is to dwarf as genius is to ?,0.296162,0.428089
3,arc is to circle as line is to ?,0.187692,0.428753
4,French is to France as Dutch is to ?,0.419229,0.604468
5,man is to woman as king is to ?,0.568557,0.553245
6,water is to ice as liquid is to ?,0.654647,0.450004
7,bad is to good as sad is to ?,0.448851,0.440382
8,nurse is to hospital as teacher is to ?,0.532657,0.482898
9,usa is to pizza as japan is to ?,0.011866,0.576351
10,human is to house as dog is to ?,0.194760,0.423166


1.3 (d)
I think I got it right for the most parts, some parts that i got lower similarity is when the word the model know that has highest similiarity just doest makes it and for this i blame on the data used to train the model. for exmple usa is to pizza as japan is to dishes?	poverty is to wealth as sickness is to impiious?	video is to cassette as computer is to peripherals?	

in fact, for analogy 7, i did better than the vector math, i gave the answer 'happy' and got similirity score of 0.449, which is higher than the score of 'glory' it computed using 'sad - bad + good' which is 0.440382

# Task 2.1 - UTK Dataset Analysis

## 2.1 (c) — Largest and least representation

Based on **9,778** images in `crop_part1`:

**Age**
- **Largest:** 0–20 (4,267 images, ~43.6%)
- **Least:** 81–116 (346 images, ~3.5%)

**Gender**
- **Largest:** 1 / female (5,406 images, ~55.3%)
- **Least:** 0 / male (4,372 images, ~44.7%)

**Race**
- **Largest:** 0 / White (5,265 images, ~53.8%)
- **Least:** 1 / Black (405 images, ~4.1%)

The imbalance is strongest for **race** (White vs. Black is roughly 13:1) and **age** (0–20 vs. 81–116 is roughly 12:1). Gender is more balanced, though females still outnumber males.

## 2.1 (d) — Who is impacted most if a model is trained on this data?

If a face-analysis model (e.g., age, gender, or race estimation) is trained on this UTK subset, **underrepresented groups will be harmed the most**. Our counts show severe skew: White subjects (~54%) dominate while Black subjects make up only ~4%; young people (0–20, ~44%) far outnumber the elderly (81–116, ~3.5%). Models learn best from groups with many examples and perform poorly on groups with few — a form of **representation bias** discussed in the course (and related to ideas in *Weapons of Math Destruction* about feedback loops and unfair impact on marginalized populations).

**Black individuals** and **elderly individuals (81–116)** would likely face the highest error rates. With only 405 Black faces and 346 elderly faces, the model sees far less variation in lighting, pose, and appearance for those groups, so predictions for them will be less reliable. **“Others”** (race 4) and **Indian** subjects are also underrepresented relative to White. Even though females are the majority, male faces (~45%) are still substantial; the race and age gaps are the more alarming imbalances.

**Real-world harm example:** A hiring or access-control system that uses age/gender/race estimation trained on this data could **misclassify older Black users** more often — for example, underestimating the age of a Black job applicant or failing facial verification at a security checkpoint. That could lead to false rejections, extra manual review, or denial of service. Because errors concentrate on already underrepresented groups, the system would **amplify existing inequality** rather than treat all demographics equally — the kind of disproportionate harm ethical AI analysis warns against when training data does not reflect the population the system will serve.

In [26]:
# 2.1 (a) - frequency counts by age group, gender, and race

import os
import pandas as pd

crop_folder = "/Users/lukecheng/OMSCS/CS6603/CS6603_2026_Summer/Project_AI_ML_Part_2/crop_part1"

rows = []
skipped = []
for fname in os.listdir(crop_folder):
    if not fname.endswith(".jpg"):
        continue
    parts = fname.split("_")
    if len(parts) < 4:
        skipped.append(fname)
        continue
    try:
        age, gender, race = int(parts[0]), int(parts[1]), int(parts[2])
    except ValueError:
        skipped.append(fname)
        continue
    rows.append({"age": age, "gender": gender, "race": race})

utk_df = pd.DataFrame(rows)
print(f"Total images: {len(utk_df)}")
if skipped:
    print(f"Skipped {len(skipped)} files with unexpected filename format: {skipped}\n")
else:
    print()

# Age groups: (0-20), (21-40), (41-60), (61-80), (81-116)
def age_group(age):
    if age <= 20:
        return "0-20"
    if age <= 40:
        return "21-40"
    if age <= 60:
        return "41-60"
    if age <= 80:
        return "61-80"
    return "81-116"

utk_df["age_group"] = utk_df["age"].apply(age_group)

age_order = ["0-20", "21-40", "41-60", "61-80", "81-116"]
age_counts = utk_df["age_group"].value_counts().reindex(age_order)
print("Age group frequencies:")
print(age_counts)
print()

gender_labels = {0: "0 (male)", 1: "1 (female)"}
gender_counts = utk_df["gender"].value_counts().sort_index()
gender_counts.index = gender_counts.index.map(gender_labels)
print("Gender frequencies:")
print(gender_counts)
print()

race_labels = {
    0: "0 (White)",
    1: "1 (Black)",
    2: "2 (Asian)",
    3: "3 (Indian)",
    4: "4 (Others)",
}
race_counts = utk_df["race"].value_counts().sort_index()
race_counts.index = race_counts.index.map(race_labels)
print("Race frequencies:")
print(race_counts)

Total images: 9778
Skipped 2 files with unexpected filename format: ['61_1_20170109142408075.jpg.chip.jpg', '61_3_20170109150557335.jpg.chip.jpg']

Age group frequencies:
age_group
0-20      4267
21-40     2533
41-60     1665
61-80      967
81-116     346
Name: count, dtype: int64

Gender frequencies:
gender
0 (male)      4372
1 (female)    5406
Name: count, dtype: int64

Race frequencies:
race
0 (White)     5265
1 (Black)      405
2 (Asian)     1553
3 (Indian)    1452
4 (Others)    1103
Name: count, dtype: int64


In [34]:
# 2.1 (b) - cross-tabulation table (matches assignment example format)
# Rows = gender + race; Columns = age groups; Values = raw counts

age_cols = ["0-20", "21-40", "41-60", "61-80", "81-116"]

def crosstab_with_total(df, row_col, row_order, row_labels):
    ct = pd.crosstab(df[row_col], df["age_group"])
    ct = ct.reindex(index=row_order, columns=age_cols, fill_value=0).astype(int)
    ct.index = [row_labels[i] for i in row_order]
    ct["Total"] = ct.sum(axis=1)
    return ct

gender_table = crosstab_with_total(
    utk_df, "gender", [1, 0], {1: "Female", 0: "Male"}
)

race_table = crosstab_with_total(
    utk_df, "race", [1, 0, 2, 3, 4],
    {1: "Black", 0: "White", 2: "Asian", 3: "Indian", 4: "Others"}
)

grand_total = utk_df["age_group"].value_counts().reindex(age_cols, fill_value=0).astype(int)
grand_total["Total"] = grand_total.sum()
grand_total_row = pd.DataFrame([grand_total.values], columns=age_cols + ["Total"], index=["Total"])

df_2_1b = pd.concat([gender_table, race_table, grand_total_row])
df_2_1b.index.name = "Age group"

print("Gender by age group:")
print(gender_table)
print("\nRace by age group:")
print(race_table)
print("\nCombined table (for report):")
df_2_1b

Gender by age group:
age_group  0-20  21-40  41-60  61-80  81-116  Total
Female     2326   1632    751    465     232   5406
Male       1941    901    914    502     114   4372

Race by age group:
age_group  0-20  21-40  41-60  61-80  81-116  Total
Black       160    100     75     55      15    405
White      1931   1034   1252    793     255   5265
Asian      1017    349     88     47      52   1553
Indian      607    598    162     63      22   1452
Others      552    452     88      9       2   1103

Combined table (for report):


,0-20,21-40,41-60,61-80,81-116,Total
Age group,,,,,,
Female,2326,1632,751,465,232,5406
Male,1941,901,914,502,114,4372
Black,160,100,75,55,15,405
White,1931,1034,1252,793,255,5265
Asian,1017,349,88,47,52,1553
Indian,607,598,162,63,22,1452
Others,552,452,88,9,2,1103
Total,4267,2533,1665,967,346,9778


## 2.1 (c) — Largest and least representation
Based on 9,778 images in `crop_part1`:
Age
- Largest: 0–20 (4,267 images, ~43.6%)
- Least:*81–116 (346 images, ~3.5%)

Gender
- Largest: 1 / female (5,406 images, ~55.3%)
- Least: 0 / male (4,372 images, ~44.7%)

Race
- Largest: 0 / White (5,265 images, ~53.8%)
- Least: 1 / Black (405 images, ~4.1%)

2.1 (d)

## 2.1 (d) — Who is impacted most if a model is trained on this data?
If a face-analysis model (e.g., age, gender, or race estimation) is trained on this UTK subset, **underrepresented groups will be harmed the most**. Our counts show severe skew: White subjects (~54%) dominate while Black subjects make up only ~4%; young people (0–20, ~44%) far outnumber the elderly (81–116, ~3.5%). Models learn best from groups with many examples and perform poorly on groups with few — a form of **representation bias** discussed in the course (and related to ideas in *Weapons of Math Destruction* about feedback loops and unfair impact on marginalized populations).

**Black individuals** and **elderly individuals (81–116)** would likely face the highest error rates. With only 405 Black faces and 346 elderly faces, the model sees far less variation in lighting, pose, and appearance for those groups, so predictions for them will be less reliable. **“Others”** (race 4) and **Indian** subjects are also underrepresented relative to White. Even though females are the majority, male faces (~45%) are still substantial; the race and age gaps are the more alarming imbalances.

**Real-world harm example:** A hiring or access-control system that uses age/gender/race estimation trained on this data could **misclassify older Black users** more often — for example, underestimating the age of a Black job applicant or failing facial verification at a security checkpoint. That could lead to false rejections, extra manual review, or denial of service. Because errors concentrate on already underrepresented groups, the system would **amplify existing inequality** rather than treat all demographics equally — the kind of disproportionate harm ethical AI analysis warns against when training data does not reflect the population the system will serve.